pkg/entities/set.go

```diff
@@ -58,22 +59,27 @@ type set struct {
 	isDecoding bool
 }
 
-func NewSet(setType ContentType, templateID uint16, isDecoding bool) Set {
-	set := &set{
+func NewSet(setType ContentType, isDecoding bool) Set {
+	return &set{
 		buffer:     &bytes.Buffer{},
 		setType:    setType,
 		records:    make([]Record, 0),
 		isDecoding: isDecoding,
 	}
-	if !isDecoding {
+}
+
+func (s *set) PrepareSet(templateID uint16) error {
+	if !s.isDecoding {
 		// Create the set header and append it when encoding
-		set.createHeader(setType, templateID)
+		return s.createHeader(s.setType, templateID)
 	}
-	return set
+	return nil
 }
 
-func (s *set) GetBuffLen() int {
-	return s.buffer.Len()
+func (s *set) ResetSet() {
+	s.buffer.Reset()
+	s.setType = Undefined
+	s.records = s.records[:0]
 }
```

Comment:
Since we will probably encounter `Undefined` for setType, we should check whether setType is `Undefined` in AddRecords function to avoid adding record to invalid set.